# Solution template for Task 1
This file is a solution template for the Task 1 of the Quantium Virtual Internship.
It will walk you through the analysis, providing the scaffolding for your solution
with gaps left for you to fill in yourself.
Look for comments that say "over to you" for places where you need to add your own
code!
Often, there will be hints about what to do or what function to use in the text
leading up to a code block - if you need a bit of extra help on how to use a
function, the internet has many excellent resources on R coding, which you can find
using your favourite search engine.
## Load required libraries and datasets
Note that you will need to install these libraries if you have never used these
before.
```{r 0 Load Libraries, results = 'hide'}
#### Example code to install packages
#install.packages("data.table")
#### Load required libraries
library(data.table)
library(ggplot2)
library(ggmosaic)
library(readr)
#### Point the filePath to where you have downloaded the datasets to and
#### assign the data files to data.tables
# over to you! fill in the path to your working directory. If you are on a Windows
machine, you will need to use forward slashes (/) instead of backshashes (\)
filePath <- ""
transactionData <- fread(paste0(filePath,"QVI_transaction_data.csv"))
customerData <- fread(paste0(filePath,"QVI_purchase_behaviour.csv"))
```

In [1]:
# To install packages
# install.packages("data.table")
# install.packages("ggplot2")
install.packages("ggmosaic")
# install.packages("readr")

# Load required libraries
library(data.table)
library(ggplot2)
library(ggmosaic)
library(readr)
library(plyr)
library(dplyr)

# Point the filePath to where you have downloaded the datasets to and
# assign the data files to data.tables
# over to you! fill in the path to your working directory. If you are on a Windows
# machine, you will need to use forward slashes (/) instead of backshashes (\)
filePath <- "./"
transactionData <- fread(paste0(filePath,"QVI_transaction_data.csv"))
customerData <- fread(paste0(filePath,"QVI_purchase_behaviour.csv"))

Retrieving 'https://packagemanager.rstudio.com/all/__linux__/focal/latest/src/contrib/ggmosaic_0.3.3.tar.gz' ...
	OK [downloaded 766 Kb in 0.2 secs]
Retrieving 'https://packagemanager.rstudio.com/all/__linux__/focal/latest/src/contrib/productplots_0.1.1.tar.gz' ...
	OK [downloaded 288.1 Kb in 0.2 secs]
Installing productplots [0.1.1] ...
	OK [installed binary]
Moving productplots [0.1.1] into the cache ...
	OK [moved to cache in 0.43 milliseconds]
Installing ggmosaic [0.3.3] ...
	OK [installed binary]
Moving ggmosaic [0.3.3] into the cache ...
	OK [moved to cache in 0.36 milliseconds]



Attaching package: ‘dplyr’


The following objects are masked from ‘package:plyr’:

    arrange, count, desc, failwith, id, mutate, rename, summarise,
    summarize


The following objects are masked from ‘package:data.table’:

    between, first, last


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union





Attaching package: ‘dplyr’


The following objects are masked from ‘package:plyr’:

    arrange, count, desc, failwith, id, mutate, rename, summarise,
    summarize


The following objects are masked from ‘package:data.table’:

    between, first, last


The following objects are masked from ‘package:stats’:

    filter, lag





Attaching package: ‘dplyr’


The following objects are masked from ‘package:plyr’:

    arrange, count, desc, failwith, id, mutate, rename, summarise,
    summarize


The following objects are masked from ‘package:data.table’:

    between, first, last





Attaching package: ‘dplyr’


The following objects are masked from ‘package:plyr’:

    arrange, count, desc, failwith, id, mutate, rename, summarise,
    summarize





Attaching package: ‘dplyr’




## Exploratory data analysis
The first step in any analysis is to first understand the data. Let's take a look
at each of the datasets provided.
### Examining transaction data
We can use `str()` to look at the format of each column and see a sample of the
data. As we have read in the dataset as a `data.table` object, we can also run
`transactionData` in the console to see a sample of the data or use
`head(transactionData)` to look at the first 10 rows.

Let's check if columns we would expect to be numeric are in numeric form and date
columns are in date format.
```{r Examining transaction data}
#### Examine transaction data
# Over to you! Examine the data using one or more of the methods described above.
```

In [2]:
str(transactionData)

Classes ‘data.table’ and 'data.frame':	264836 obs. of  8 variables:
 $ DATE          : int  43390 43599 43605 43329 43330 43604 43601 43601 43332 43330 ...
 $ STORE_NBR     : int  1 1 1 2 2 4 4 4 5 7 ...
 $ LYLTY_CARD_NBR: int  1000 1307 1343 2373 2426 4074 4149 4196 5026 7150 ...
 $ TXN_ID        : int  1 348 383 974 1038 2982 3333 3539 4525 6900 ...
 $ PROD_NBR      : int  5 66 61 69 108 57 16 24 42 52 ...
 $ PROD_NAME     : chr  "Natural Chip        Compny SeaSalt175g" "CCs Nacho Cheese    175g" "Smiths Crinkle Cut  Chips Chicken 170g" "Smiths Chip Thinly  S/Cream&Onion 175g" ...
 $ PROD_QTY      : int  2 3 2 5 3 1 1 1 1 2 ...
 $ TOT_SALES     : num  6 6.3 2.9 15 13.8 5.1 5.7 3.6 3.9 7.2 ...
 - attr(*, ".internal.selfref")=<externalptr> 


We can see that the date column is in an integer format. Let's change this to a
date format.
```{r Convert DATE to date format}
#### Convert DATE column to a date format
#### A quick search online tells us that CSV and Excel integer dates begin on 30
Dec 1899
transactionData$DATE <- as.Date(transactionData$DATE, origin = "1899-12-30")
```

In [3]:
# Convert DATE column to a date format
# A quick search online tells us that CSV and Excel integer dates begin on 30 Dec 1899
transactionData$DATE <- as.Date(transactionData$DATE, origin = "1899-12-30")
str(transactionData)

Classes ‘data.table’ and 'data.frame':	264836 obs. of  8 variables:
 $ DATE          : Date, format: "2018-10-17" "2019-05-14" ...
 $ STORE_NBR     : int  1 1 1 2 2 4 4 4 5 7 ...
 $ LYLTY_CARD_NBR: int  1000 1307 1343 2373 2426 4074 4149 4196 5026 7150 ...
 $ TXN_ID        : int  1 348 383 974 1038 2982 3333 3539 4525 6900 ...
 $ PROD_NBR      : int  5 66 61 69 108 57 16 24 42 52 ...
 $ PROD_NAME     : chr  "Natural Chip        Compny SeaSalt175g" "CCs Nacho Cheese    175g" "Smiths Crinkle Cut  Chips Chicken 170g" "Smiths Chip Thinly  S/Cream&Onion 175g" ...
 $ PROD_QTY      : int  2 3 2 5 3 1 1 1 1 2 ...
 $ TOT_SALES     : num  6 6.3 2.9 15 13.8 5.1 5.7 3.6 3.9 7.2 ...
 - attr(*, ".internal.selfref")=<externalptr> 


We should check that we are looking at the right products by examining PROD_NAME.
```{r Summary of PROD_NAME}
#### Examine PROD_NAME
# Over to you! Generate a summary of the PROD_NAME column.
```

In [4]:
summary(transactionData)
summary(transactionData$PROD_NAME)

      DATE              STORE_NBR     LYLTY_CARD_NBR        TXN_ID       
 Min.   :2018-07-01   Min.   :  1.0   Min.   :   1000   Min.   :      1  
 1st Qu.:2018-09-30   1st Qu.: 70.0   1st Qu.:  70021   1st Qu.:  67602  
 Median :2018-12-30   Median :130.0   Median : 130358   Median : 135138  
 Mean   :2018-12-30   Mean   :135.1   Mean   : 135550   Mean   : 135158  
 3rd Qu.:2019-03-31   3rd Qu.:203.0   3rd Qu.: 203094   3rd Qu.: 202701  
 Max.   :2019-06-30   Max.   :272.0   Max.   :2373711   Max.   :2415841  
    PROD_NBR       PROD_NAME            PROD_QTY         TOT_SALES      
 Min.   :  1.00   Length:264836      Min.   :  1.000   Min.   :  1.500  
 1st Qu.: 28.00   Class :character   1st Qu.:  2.000   1st Qu.:  5.400  
 Median : 56.00   Mode  :character   Median :  2.000   Median :  7.400  
 Mean   : 56.58                      Mean   :  1.907   Mean   :  7.304  
 3rd Qu.: 85.00                      3rd Qu.:  2.000   3rd Qu.:  9.200  
 Max.   :114.00                      Max.   

   Length     Class      Mode 
   264836 character character 

Looks like we are definitely looking at potato chips but how can we check that
these are all chips? We can do some basic text analysis by summarising the
individual words in the product name.
```{r Further examine PROD_NAME}
#### Examine the words in PROD_NAME to see if there are any incorrect entries
#### such as products that are not chips
productWords <- data.table(unlist(strsplit(unique(transactionData[, PROD_NAME]), "
")))
setnames(productWords, 'words')
```

In [5]:
productWords <- data.table(unlist(strsplit(unique(transactionData[, PROD_NAME]), "
")))
setnames(productWords, 'words')
productWords

words                                   
1   Natural Chip        Compny SeaSalt175g  
2   CCs Nacho Cheese    175g                
3   Smiths Crinkle Cut  Chips Chicken 170g  
4   Smiths Chip Thinly  S/Cream&Onion 175g  
5   Kettle Tortilla ChpsHny&Jlpno Chili 150g
6   Old El Paso Salsa   Dip Tomato Mild 300g
7   Smiths Crinkle Chips Salt & Vinegar 330g
8   Grain Waves         Sweet Chilli 210g   
9   Doritos Corn Chip Mexican Jalapeno 150g 
10  Grain Waves Sour    Cream&Chives 210G   
11  Kettle Sensations   Siracha Lime 150g   
12  Twisties Cheese     270g                
13  WW Crinkle Cut      Chicken 175g        
14  Thins Chips Light&  Tangy 175g          
15  CCs Original 175g                       
16  Burger Rings 220g                       
17  NCC Sour Cream &    Garden Chives 175g  
18  Doritos Corn Chip Southern Chicken 150g 
19  Cheezels Cheese Box 125g                
20  Smiths Crinkle      Original 330g       
21  Infzns Crn Crnchers Tangy Gcamole 110g  
22  Kettle Sea Salt     And Vinegar 175g    
23  Smiths Chip Thinly  Cut Original 175g   
24  Kettle Original 175g                    
25  Red Rock Deli Thai  Chilli&Lime 150g    
26  Pringles Sthrn FriedChicken 134g        
27  Pringles Sweet&Spcy BBQ 134g            
28  Red Rock Deli SR    Salsa & Mzzrlla 150g
29  Thins Chips         Originl saltd 175g  
30  Red Rock Deli Sp    Salt & Truffle 150G 
⋮   ⋮                                       
85  Tostitos Splash Of  Lime 175g           
86  Woolworths Medium   Salsa 300g          
87  Kettle Tortilla ChpsBtroot&Ricotta 150g 
88  CCs Tasty Cheese    175g                
89  Woolworths Cheese   Rings 190g          
90  Tostitos Smoked     Chipotle 175g       
91  Pringles Barbeque   134g                
92  WW Supreme Cheese   Corn Chips 200g     
93  Pringles Mystery    Flavour 134g        
94  Tyrrells Crisps     Ched & Chives 165g  
95  Snbts Whlgrn Crisps Cheddr&Mstrd 90g    
96  Cheetos Chs & Bacon Balls 190g          
97  Pringles Slt Vingar 134g                
98  Infuzions SourCream&Herbs Veg Strws 110g
99  Kettle Tortilla ChpsFeta&Garlic 150g    
100 Infuzions Mango     Chutny Papadums 70g 
101 RRD Steak &         Chimuchurri 150g    
102 RRD Honey Soy       Chicken 165g        
103 Sunbites Whlegrn    Crisps Frch/Onin 90g
104 RRD Salt & Vinegar  165g                
105 Doritos Cheese      Supreme 330g        
106 Smiths Crinkle Cut  Snag&Sauce 150g     
107 WW Sour Cream &OnionStacked Chips 160g  
108 RRD Lime & Pepper   165g                
109 Natural ChipCo Sea  Salt & Vinegr 175g  
110 Red Rock Deli Chikn&Garlic Aioli 150g   
111 RRD SR Slow Rst     Pork Belly 150g     
112 RRD Pc Sea Salt     165g                
113 Smith Crinkle Cut   Bolognese 150g      
114 Doritos Salsa Mild  300g

As we are only interested in words that will tell us if the product is chips or
not, let's remove all words with digits and special characters such as '&' from our
set of product words. We can do this using `grepl()`.
```{r}
# Over to you! Remove digits, and special characters, and then sort the distinct
words by frequency of occurrence.
#### Removing digits
#### Removing special characters
#### Let's look at the most common words by counting the number of times a word
#### appears and sorting them by this frequency in order of highest 
#### to lowest frequency
```

In [6]:
sort(productWords$words)
#### Removing digits
#### Removing special characters
#### Let's look at the most common words by counting the number of times a word 
#### appears and sorting them by this frequency in order of highest 
#### to lowest frequency

[1] "Burger Rings 220g"                       
  [2] "CCs Nacho Cheese    175g"                
  [3] "CCs Original 175g"                       
  [4] "CCs Tasty Cheese    175g"                
  [5] "Cheetos Chs & Bacon Balls 190g"          
  [6] "Cheetos Puffs 165g"                      
  [7] "Cheezels Cheese 330g"                    
  [8] "Cheezels Cheese Box 125g"                
  [9] "Cobs Popd Sea Salt  Chips 110g"          
 [10] "Cobs Popd Sour Crm  &Chives Chips 110g"  
 [11] "Cobs Popd Swt/Chlli &Sr/Cream Chips 110g"
 [12] "Dorito Corn Chp     Supreme 380g"        
 [13] "Doritos Cheese      Supreme 330g"        
 [14] "Doritos Corn Chip Mexican Jalapeno 150g" 
 [15] "Doritos Corn Chip Southern Chicken 150g" 
 [16] "Doritos Corn Chips  Cheese Supreme 170g" 
 [17] "Doritos Corn Chips  Nacho Cheese 170g"   
 [18] "Doritos Corn Chips  Original 170g"       
 [19] "Doritos Mexicana    170g"                
 [20] "Doritos Salsa       Medium 300g"         
 [21] "Doritos Salsa Mild  300g"                
 [22] "French Fries Potato Chips 175g"          
 [23] "Grain Waves         Sweet Chilli 210g"   
 [24] "Grain Waves Sour    Cream&Chives 210G"   
 [25] "GrnWves Plus Btroot & Chilli Jam 180g"   
 [26] "Infuzions BBQ Rib   Prawn Crackers 110g" 
 [27] "Infuzions Mango     Chutny Papadums 70g" 
 [28] "Infuzions SourCream&Herbs Veg Strws 110g"
 [29] "Infuzions Thai SweetChili PotatoMix 110g"
 [30] "Infzns Crn Crnchers Tangy Gcamole 110g"  
 [31] "Kettle 135g Swt Pot Sea Salt"            
 [32] "Kettle Chilli 175g"                      
 [33] "Kettle Honey Soy    Chicken 175g"        
 [34] "Kettle Mozzarella   Basil & Pesto 175g"  
 [35] "Kettle Original 175g"                    
 [36] "Kettle Sea Salt     And Vinegar 175g"    
 [37] "Kettle Sensations   BBQ&Maple 150g"      
 [38] "Kettle Sensations   Camembert & Fig 150g"
 [39] "Kettle Sensations   Siracha Lime 150g"   
 [40] "Kettle Sweet Chilli And Sour Cream 175g" 
 [41] "Kettle Tortilla ChpsBtroot&Ricotta 150g" 
 [42] "Kettle Tortilla ChpsFeta&Garlic 150g"    
 [43] "Kettle Tortilla ChpsHny&Jlpno Chili 150g"
 [44] "Natural Chip        Compny SeaSalt175g"  
 [45] "Natural Chip Co     Tmato Hrb&Spce 175g" 
 [46] "Natural ChipCo      Hony Soy Chckn175g"  
 [47] "Natural ChipCo Sea  Salt & Vinegr 175g"  
 [48] "NCC Sour Cream &    Garden Chives 175g"  
 [49] "Old El Paso Salsa   Dip Chnky Tom Ht300g"
 [50] "Old El Paso Salsa   Dip Tomato Med 300g" 
 [51] "Old El Paso Salsa   Dip Tomato Mild 300g"
 [52] "Pringles Barbeque   134g"                
 [53] "Pringles Chicken    Salt Crips 134g"     
 [54] "Pringles Mystery    Flavour 134g"        
 [55] "Pringles Original   Crisps 134g"         
 [56] "Pringles Slt Vingar 134g"                
 [57] "Pringles SourCream  Onion 134g"          
 [58] "Pringles Sthrn FriedChicken 134g"        
 [59] "Pringles Sweet&Spcy BBQ 134g"            
 [60] "Red Rock Deli Chikn&Garlic Aioli 150g"   
 [61] "Red Rock Deli Sp    Salt & Truffle 150G" 
 [62] "Red Rock Deli SR    Salsa & Mzzrlla 150g"
 [63] "Red Rock Deli Thai  Chilli&Lime 150g"    
 [64] "RRD Chilli&         Coconut 150g"        
 [65] "RRD Honey Soy       Chicken 165g"        
 [66] "RRD Lime & Pepper   165g"                
 [67] "RRD Pc Sea Salt     165g"                
 [68] "RRD Salt & Vinegar  165g"                
 [69] "RRD SR Slow Rst     Pork Belly 150g"     
 [70] "RRD Steak &         Chimuchurri 150g"    
 [71] "RRD Sweet Chilli &  Sour Cream 165g"     
 [72] "Smith Crinkle Cut   Bolognese 150g"      
 [73] "Smith Crinkle Cut   Mac N Cheese 150g"   
 [74] "Smiths Chip Thinly  Cut Original 175g"   
 [75] "Smiths Chip Thinly  CutSalt/Vinegr175g"  
 [76] "Smiths Chip Thinly  S/Cream&Onion 175g"  
 [77] "Smiths Crinkle      Original 330g"       
 [78] "Smiths Crinkle Chips Salt & Vinegar 330g"
 [79] "Smiths Crinkle Cut  Chips Barbecue 170g" 
 [80] "Smiths Crinkle Cut  Chips Chicken 170g"  
 [81] "Smiths Crinkle Cut  Chips Chs&Onion170g" 
 [82] "Smiths Crinkle Cut  Chips 

There are salsa products in the dataset but we are only interested in the chips
category, so let's remove these.
```{r}
#### Remove salsa products
transactionData[, SALSA := grepl("salsa", tolower(PROD_NAME))]
transactionData <- transactionData[SALSA == FALSE, ][, SALSA := NULL]
```

In [7]:
#### Remove salsa products
transactionData[, SALSA := grepl("salsa", tolower(PROD_NAME))]
transactionData <- transactionData[SALSA == FALSE, ][, SALSA := NULL]
print(transactionData)

              DATE STORE_NBR LYLTY_CARD_NBR TXN_ID PROD_NBR
     1: 2018-10-17         1           1000      1        5
     2: 2019-05-14         1           1307    348       66
     3: 2019-05-20         1           1343    383       61
     4: 2018-08-17         2           2373    974       69
     5: 2018-08-18         2           2426   1038      108
    ---                                                    
246738: 2019-03-09       272         272319 270088       89
246739: 2018-08-13       272         272358 270154       74
246740: 2018-11-06       272         272379 270187       51
246741: 2018-12-27       272         272379 270188       42
246742: 2018-09-22       272         272380 270189       74
                                       PROD_NAME PROD_QTY TOT_SALES
     1:   Natural Chip        Compny SeaSalt175g        2       6.0
     2:                 CCs Nacho Cheese    175g        3       6.3
     3:   Smiths Crinkle Cut  Chips Chicken 170g        2       2.9
     4: 

Next, we can use `summary()` to check summary statistics such as mean, min and max
values for each feature to see if there are any obvious outliers in the data and if
there are any nulls in any of the columns (`NA's : number of nulls` will appear in
the output if there are any nulls).
```{r initial summary}
#### Summarise the data to check for nulls and possible outliers
# Over to you!
```

In [8]:
summary(transactionData)

      DATE              STORE_NBR     LYLTY_CARD_NBR        TXN_ID       
 Min.   :2018-07-01   Min.   :  1.0   Min.   :   1000   Min.   :      1  
 1st Qu.:2018-09-30   1st Qu.: 70.0   1st Qu.:  70015   1st Qu.:  67569  
 Median :2018-12-30   Median :130.0   Median : 130367   Median : 135183  
 Mean   :2018-12-30   Mean   :135.1   Mean   : 135531   Mean   : 135131  
 3rd Qu.:2019-03-31   3rd Qu.:203.0   3rd Qu.: 203084   3rd Qu.: 202654  
 Max.   :2019-06-30   Max.   :272.0   Max.   :2373711   Max.   :2415841  
    PROD_NBR       PROD_NAME            PROD_QTY         TOT_SALES      
 Min.   :  1.00   Length:246742      Min.   :  1.000   Min.   :  1.700  
 1st Qu.: 26.00   Class :character   1st Qu.:  2.000   1st Qu.:  5.800  
 Median : 53.00   Mode  :character   Median :  2.000   Median :  7.400  
 Mean   : 56.35                      Mean   :  1.908   Mean   :  7.321  
 3rd Qu.: 87.00                      3rd Qu.:  2.000   3rd Qu.:  8.800  
 Max.   :114.00                      Max.   

There are no nulls in the columns but product quantity appears to have an outlier
which we should investigate further. Let's investigate further the case where 200
packets of chips are bought in one transaction.
```{r }
#### Filter the dataset to find the outlier
# Over to you! Use a filter to examine the transactions in question.
```

In [9]:
filter(transactionData, transactionData$PROD_QTY == 200)

DATE       STORE_NBR LYLTY_CARD_NBR TXN_ID PROD_NBR
1 2018-08-19 226       226000         226201 4       
2 2019-05-20 226       226000         226210 4       
  PROD_NAME                        PROD_QTY TOT_SALES
1 Dorito Corn Chp     Supreme 380g 200      650      
2 Dorito Corn Chp     Supreme 380g 200      650

There are two transactions where 200 packets of chips are bought in one transaction
and both of these transactions were by the same customer.
```{r}
#### Let's see if the customer has had other transactions
# Over to you! Use a filter to see what other transactions that customer made.
```

In [10]:
filter(transactionData, transactionData$LYLTY_CARD_NBR == 226000)

DATE       STORE_NBR LYLTY_CARD_NBR TXN_ID PROD_NBR
1 2018-08-19 226       226000         226201 4       
2 2019-05-20 226       226000         226210 4       
  PROD_NAME                        PROD_QTY TOT_SALES
1 Dorito Corn Chp     Supreme 380g 200      650      
2 Dorito Corn Chp     Supreme 380g 200      650

It looks like this customer has only had the two transactions over the year and is
not an ordinary retail customer. The customer might be buying chips for commercial
purposes instead. We'll remove this loyalty card number from further analysis.
```{r}
#### Filter out the customer based on the loyalty card number
# Over to you!
#### Re-examine transaction data
# Over to you!
```

In [11]:
transactionData[, COMM := grepl(226000, LYLTY_CARD_NBR)]
transactionData <- transactionData[COMM == FALSE, ][, COMM := NULL]
summary(transactionData)                                

      DATE              STORE_NBR     LYLTY_CARD_NBR        TXN_ID       
 Min.   :2018-07-01   Min.   :  1.0   Min.   :   1000   Min.   :      1  
 1st Qu.:2018-09-30   1st Qu.: 70.0   1st Qu.:  70015   1st Qu.:  67569  
 Median :2018-12-30   Median :130.0   Median : 130367   Median : 135182  
 Mean   :2018-12-30   Mean   :135.1   Mean   : 135530   Mean   : 135130  
 3rd Qu.:2019-03-31   3rd Qu.:203.0   3rd Qu.: 203083   3rd Qu.: 202652  
 Max.   :2019-06-30   Max.   :272.0   Max.   :2373711   Max.   :2415841  
    PROD_NBR       PROD_NAME            PROD_QTY       TOT_SALES     
 Min.   :  1.00   Length:246740      Min.   :1.000   Min.   : 1.700  
 1st Qu.: 26.00   Class :character   1st Qu.:2.000   1st Qu.: 5.800  
 Median : 53.00   Mode  :character   Median :2.000   Median : 7.400  
 Mean   : 56.35                      Mean   :1.906   Mean   : 7.316  
 3rd Qu.: 87.00                      3rd Qu.:2.000   3rd Qu.: 8.800  
 Max.   :114.00                      Max.   :5.000   Max.   :2

That's better. Now, let's look at the number of transaction lines over time to see
if there are any obvious data issues such as missing data.
```{r}
#### Count the number of transactions by date
# Over to you! Create a summary of transaction count by date.
```

In [12]:
transactions_by_date <- count(transactionData, DATE)
transactions_by_date

DATE       n  
1   2018-07-01 663
2   2018-07-02 650
3   2018-07-03 674
4   2018-07-04 669
5   2018-07-05 660
6   2018-07-06 711
7   2018-07-07 695
8   2018-07-08 653
9   2018-07-09 692
10  2018-07-10 650
11  2018-07-11 701
12  2018-07-12 717
13  2018-07-13 727
14  2018-07-14 661
15  2018-07-15 712
16  2018-07-16 678
17  2018-07-17 694
18  2018-07-18 689
19  2018-07-19 637
20  2018-07-20 684
21  2018-07-21 683
22  2018-07-22 673
23  2018-07-23 673
24  2018-07-24 648
25  2018-07-25 674
26  2018-07-26 672
27  2018-07-27 697
28  2018-07-28 640
29  2018-07-29 659
30  2018-07-30 692
⋮   ⋮          ⋮  
335 2019-06-01 682
336 2019-06-02 662
337 2019-06-03 656
338 2019-06-04 637
339 2019-06-05 680
340 2019-06-06 700
341 2019-06-07 762
342 2019-06-08 699
343 2019-06-09 718
344 2019-06-10 676
345 2019-06-11 634
346 2019-06-12 709
347 2019-06-13 607
348 2019-06-14 743
349 2019-06-15 724
350 2019-06-16 690
351 2019-06-17 658
352 2019-06-18 639
353 2019-06-19 662
354 2019-06-20 698
355 2019-06-21 716
356 2019-06-22 643
357 2019-06-23 653
358 2019-06-24 612
359 2019-06-25 696
360 2019-06-26 657
361 2019-06-27 669
362 2019-06-28 673
363 2019-06-29 703
364 2019-06-30 704

There's only 364 rows, meaning only 364 dates which indicates a missing date. Let's
create a sequence of dates from 1 Jul 2018 to 30 Jun 2019 and use this to create a
chart of number of transactions over time to find the missing date.
```{r fig.align = "center"}
#### Create a sequence of dates and join this the count of transactions by date
# Over to you - create a column of dates that includes every day from 1 Jul 2018 to 30 Jun 2019, and join it onto the data to fill in the missing day.
#### Setting plot themes to format graphs
theme_set(theme_bw())
theme_update(plot.title = element_text(hjust = 0.5))
#### Plot transactions over time
ggplot(transactions_by_day, aes(x = DATE, y = N)) +
geom_line() +
labs(x = "Day", y = "Number of transactions", title = "Transactions over time") +
scale_x_date(breaks = "1 month") +
theme(axis.text.x = element_text(angle = 90, vjust = 0.5))
```

In [13]:
target_dates <- c(seq(as.Date("2018/7/1"), as.Date("2019/6/30"), 'day'))
target_dates
#merge(target_dates, transactions_by_date)

[1] "2018-07-01" "2018-07-02" "2018-07-03" "2018-07-04" "2018-07-05"
  [6] "2018-07-06" "2018-07-07" "2018-07-08" "2018-07-09" "2018-07-10"
 [11] "2018-07-11" "2018-07-12" "2018-07-13" "2018-07-14" "2018-07-15"
 [16] "2018-07-16" "2018-07-17" "2018-07-18" "2018-07-19" "2018-07-20"
 [21] "2018-07-21" "2018-07-22" "2018-07-23" "2018-07-24" "2018-07-25"
 [26] "2018-07-26" "2018-07-27" "2018-07-28" "2018-07-29" "2018-07-30"
 [31] "2018-07-31" "2018-08-01" "2018-08-02" "2018-08-03" "2018-08-04"
 [36] "2018-08-05" "2018-08-06" "2018-08-07" "2018-08-08" "2018-08-09"
 [41] "2018-08-10" "2018-08-11" "2018-08-12" "2018-08-13" "2018-08-14"
 [46] "2018-08-15" "2018-08-16" "2018-08-17" "2018-08-18" "2018-08-19"
 [51] "2018-08-20" "2018-08-21" "2018-08-22" "2018-08-23" "2018-08-24"
 [56] "2018-08-25" "2018-08-26" "2018-08-27" "2018-08-28" "2018-08-29"
 [61] "2018-08-30" "2018-08-31" "2018-09-01" "2018-09-02" "2018-09-03"
 [66] "2018-09-04" "2018-09-05" "2018-09-06" "2018-09-07" "2018-09-08"
 [71] "2018-09-09" "2018-09-10" "2018-09-11" "2018-09-12" "2018-09-13"
 [76] "2018-09-14" "2018-09-15" "2018-09-16" "2018-09-17" "2018-09-18"
 [81] "2018-09-19" "2018-09-20" "2018-09-21" "2018-09-22" "2018-09-23"
 [86] "2018-09-24" "2018-09-25" "2018-09-26" "2018-09-27" "2018-09-28"
 [91] "2018-09-29" "2018-09-30" "2018-10-01" "2018-10-02" "2018-10-03"
 [96] "2018-10-04" "2018-10-05" "2018-10-06" "2018-10-07" "2018-10-08"
[101] "2018-10-09" "2018-10-10" "2018-10-11" "2018-10-12" "2018-10-13"
[106] "2018-10-14" "2018-10-15" "2018-10-16" "2018-10-17" "2018-10-18"
[111] "2018-10-19" "2018-10-20" "2018-10-21" "2018-10-22" "2018-10-23"
[116] "2018-10-24" "2018-10-25" "2018-10-26" "2018-10-27" "2018-10-28"
[121] "2018-10-29" "2018-10-30" "2018-10-31" "2018-11-01" "2018-11-02"
[126] "2018-11-03" "2018-11-04" "2018-11-05" "2018-11-06" "2018-11-07"
[131] "2018-11-08" "2018-11-09" "2018-11-10" "2018-11-11" "2018-11-12"
[136] "2018-11-13" "2018-11-14" "2018-11-15" "2018-11-16" "2018-11-17"
[141] "2018-11-18" "2018-11-19" "2018-11-20" "2018-11-21" "2018-11-22"
[146] "2018-11-23" "2018-11-24" "2018-11-25" "2018-11-26" "2018-11-27"
[151] "2018-11-28" "2018-11-29" "2018-11-30" "2018-12-01" "2018-12-02"
[156] "2018-12-03" "2018-12-04" "2018-12-05" "2018-12-06" "2018-12-07"
[161] "2018-12-08" "2018-12-09" "2018-12-10" "2018-12-11" "2018-12-12"
[166] "2018-12-13" "2018-12-14" "2018-12-15" "2018-12-16" "2018-12-17"
[171] "2018-12-18" "2018-12-19" "2018-12-20" "2018-12-21" "2018-12-22"
[176] "2018-12-23" "2018-12-24" "2018-12-25" "2018-12-26" "2018-12-27"
[181] "2018-12-28" "2018-12-29" "2018-12-30" "2018-12-31" "2019-01-01"
[186] "2019-01-02" "2019-01-03" "2019-01-04" "2019-01-05" "2019-01-06"
[191] "2019-01-07" "2019-01-08" "2019-01-09" "2019-01-10" "2019-01-11"
[196] "2019-01-12" "2019-01-13" "2019-01-14" "2019-01-15" "2019-01-16"
[201] "2019-01-17" "2019-01-18" "2019-01-19" "2019-01-20" "2019-01-21"
[206] "2019-01-22" "2019-01-23" "2019-01-24" "2019-01-25" "2019-01-26"
[211] "2019-01-27" "2019-01-28" "2019-01-29" "2019-01-30" "2019-01-31"
[216] "2019-02-01" "2019-02-02" "2019-02-03" "2019-02-04" "2019-02-05"
[221] "2019-02-06" "2019-02-07" "2019-02-08" "2019-02-09" "2019-02-10"
[226] "2019-02-11" "2019-02-12" "2019-02-13" "2019-02-14" "2019-02-15"
[231] "2019-02-16" "2019-02-17" "2019-02-18" "2019-02-19" "2019-02-20"
[236] "2019-02-21" "2019-02-22" "2019-02-23" "2019-02-24" "2019-02-25"
[241] "2019-02-26" "2019-02-27" "2019-02-28" "2019-03-01" "2019-03-02"
[246] "2019-03-03" "2019-03-04" "2019-03-05" "2019-03-06" "2019-03-07"
[251] "2019-03-08" "2019-03-09" "2019-03-10" "2019-03-11" "2019-03-12"
[256] "2019-03-13" "2019-03-14" "2019-03-15" "2019-03-16" "2019-03-17"
[261] "2019-03-18" "2019-03-19" "2019-03-20" "2019-03-21" "2019-03-22"
[266] "2019-03-23" "2019-03-24" "2019-03-25" "2019-03-26" "2019-03-27"
[271] "2019-03-28" "2019-03-29" "2019-03-30" "2019-03-31" "2019-04-01"
[276] "2019-04-02" "2019-04-03" "2019-04-04" "2019-04-05" "2019-04-06"
[281] "2019-04-07" "2019-0